master/internal/provisioner/aws_spot.go

```diff
@@ -0,0 +1,683 @@
+// fulfilled an actual EC2 instance will be returned as an Instance. If the request has not
+// been fulfilled, a fake Instance will be returned where the InstanceId is the SpotRequestId
+// and the state is SpotRequestPendingAWS.
+func (c *awsCluster) listSpot(ctx *actor.Context) ([]*Instance, error) {
+	activeReqsInAPI, err := c.listActiveSpotInstanceRequests(ctx, false)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot describe EC2 spot requests")
+	}
+
+	// Make sure we're tracking all active reqs (to handle master restart)
+	for _, req := range activeReqsInAPI.iter() {
+		c.spot.trackedReqs.add(req)
+	}
+
+	// If there are requests that we are tracking, but didn't get returned in listActiveSpotInstanceRequests,
+	// query the API for them specifically. They could be fresh requests that aren't visible in the API yet
+	// or they could have entered an inactive state for some reason.
+	missingReqs := c.spot.trackedReqs.copy()
+	missingReqs.deleteIntersection(*activeReqsInAPI)
+
+	newOrInactiveReqs, err := c.listSpotRequestsById(ctx, missingReqs.idsAsListOfPointers(), false)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot describe EC2 spot requests")
+	}
+
+	// If any of the tracked requests failed and requires users intervention, notify the user via error
+	// logs. Also stop tracking any request that is in a terminal state.
+	reqsToNotifyUserAbout := newSetOfSpotRequests()
+	numReqsNoLongerTracked := 0
+	for _, req := range newOrInactiveReqs.iter() {
+		missingReqs.delete(req)
+		if req.State != "active" && req.State != "open" {
+			c.spot.trackedReqs.delete(req)
+			numReqsNoLongerTracked++
+		}
+		switch *req.StatusCode {
+		case
+			"bad-parameters",
+			"constraint-not-fulfillable",
+			"limit-exceeded":
+			reqsToNotifyUserAbout.add(req)
+		}
+	}
+
+	for _, req := range reqsToNotifyUserAbout.asListInChronologicalOrder() {
+		ctx.Log().
+			WithField("spot-request-status-code", *req.StatusCode).
+			WithField("spot-request-status-message", *req.StatusMessage).
+			WithField("spot-request-creation-time", req.CreationTime.String()).
+			Error("a spot request cannot be fulfilled and may require user intervention")
+	}
+
+	// Canonical log line for debugging
+	ctx.Log().
+		WithField("log-type", "listSpot.summary").
+		WithField("total-num-requests-being-tracked", c.spot.trackedReqs.numReqs()).
+		WithField("num-visible-as-active-in-api", activeReqsInAPI.numReqs()).
+		WithField("num-tracked-but-not-visible-in-aws-api", missingReqs.numReqs()).
+		WithField("num-no-longer-tracked-due-to-terminal-state", numReqsNoLongerTracked).
+		Debugf("updated the list of active spot requests being tracked. "+
+			"there are %d spot requests being tracked. the following requests "+
+			"are being tracked but aren't visible in AWS yet: %s",
+			c.spot.trackedReqs.numReqs(),
+			missingReqs.idsAsList())
+
+	// Cleanup CanceledButInstanceRunningRequests to make sure race conditions + eventual
```

Comment:
It is a cleanup phase to handle any issues around this being a distributed system with eventual consistency. It isn't intended to handle a specific issue (although the race condition called out in terminateSpot is one of the things that gets cleaned up here). 

It introduces an invariant into our system (any spot request that is cancelled while the instance is still running will be cleaned up) and protects us in case the distributed system fails in a way we haven't thought of.

Reply:
we should move the comment on the race condition from `terminateSpot` to here since `terminateSpot` is just executing what is told by the scale decider.